### PR TITLE
[RFC] Add the notion of services

### PR DESCRIPTION
**Pull request by nono**

### CODE REVIEW

File: doc/service.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+It's obvious that services should be benefit from the authentication
+mechanisms used by Cozy-Proxy. For the permissions, it's more complicated.
+Currently, they are used in cozy-data-system and are tied to the DocTypes in
+CouchDB documents. I think we should generalize permissions.
+
+For example, a service to send an email to the owner of the cozy instance can
+be useful (think of Kresus sending you an email when your balance is too low).
+This service will not manipulate CouchDB documents. But, even in absence of
+doctypes, it should be possible to put a permission for this service.
```
Comment (jsilvestre) [inline doc/service.md]: Not entirely true, you must already ask for specific permission to "send mails to user" or "send mails", and so on (or it has changed since the last time I hack that code)

**Comment by jsilvestre** [comment]
Security-wise, what are the possibility to sandbox services? Right now one service could use another by require-ing its file and do stuff with it.
I remember it was one of the topic we had not addressed when @Kloadut first tried to do something similar.

**Comment by frankrousseau** [comment]
For me services should be available at the data-system level.

File: doc/service.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+use less memory.
+
+I think to embed the services in cozy-proxy and not cozy-data-system beacuse
+cozy-data-system could become a service its-self (some day, in the long term).
+And most server-side code from cozy-home too. Maybe, we can even unify all of
+our nodejs code to run in the same processus (I'm dreaming).
+
+### 3. As express/americano middlewares
+
+It's the simple way I see to make this happen.
```
Comment (frankrousseau) [inline doc/service.md]: I would say as an express app. You can mount apps inside another app with Express.

File: doc/service.md (hunk 1 above)
Comment (nono) [inline doc/service.md]: OK, my bad, I didn't know that.

Comment (jsilvestre) [inline doc/service.md]: Your point of having permissions unrelated to doctype is still very valid anyway!

**Comment by nono** [comment]
@jsilvestre for the security, you are right, I don't see how to sandbox them. That's why I think services should be only for code reviewed by cozy, as explained in 10. (at least for the near future).

File: doc/service.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+Rationale
+---------
+
+### 1. It's some JavaScript code
+
+JS is the lowest common denominator for all the code written at Cozy.
+Technically, a service can be written in something else (CoffeeScript or
+TypeScript) but it has to be compiled simply to JavaScript.
+
+### 2. That is embedded in Cozy-Proxy
```
Comment (m4dz) [inline doc/service.md]: I can understand why host all services in cozy-proxy (it's the only exposed server behind the reverse-proxy, it's already deployed, it permits customization for other parts like Home (you can develop a custom home for your own needs if you want w/o redevelopping all stuff behind the offical one), etc) but I ask myself about the name and architecture of the proxy to do this.

Don't we need to rename it something like "cozy-services", with a smart and clever plugin API to plug services (consider auth as a builtin plugin) to it?

File: doc/service.md [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+I think to embed the services in cozy-proxy and not cozy-data-system beacuse
+cozy-data-system could become a service its-self (some day, in the long term).
+And most server-side code from cozy-home too. Maybe, we can even unify all of
+our nodejs code to run in the same processus (I'm dreaming).
+
+### 3. As express/americano apps
+
+It's the simple way I see to make this happen.
+
+### 4. And exposes a REST API in JSON
```
Comment (m4dz) [inline doc/service.md]: _note_ maybe [PSON](https://github.com/dcodeIO/PSON) or [ProtocolBuffer](https://github.com/dcodeIO/protobuf.js/wiki) may be better choices than JSON to expose data to clients.

File: doc/service.md [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+Cozy applications can be run in the browser with cozysdk, on the server,
+but there are also cozy-mobile and cozy-desktop. All of them should be able to
+use the services even if the authentication mechanisms are not the same.
+
+### 6. This API follows the [JSON api specification](http://jsonapi.org/)
+
+I have nothing against [JSON-ld](http://json-ld.org/). The important is to
+take one standard and stick to it. JSON api looks good. So, go for it!
+
```
Comment (m4dz) [inline doc/service.md]: _note_ maybe [PSON](https://github.com/dcodeIO/PSON) or [ProtocolBuffer](https://github.com/dcodeIO/protobuf.js/wiki) may be better choices than JSON to expose data to clients.

**Comment by aenario** [comment]
OK, this is way cool and I agree with almost everything except for the following :


2. I think services should belong to the data-system : 
   - the DS already has all the code and logic to deal with permissions
   - the current "services" are all in the DS.
   - the entire DS API is already accessible from the exterior through the `/ds-api/*` route
   - if we want to imagine some future architecture, it makes AMHA more sense to have a DS handling cookies as a form of auth & serving clientside apps than moving the entire DS as a service of the proxy. 
   - keeping the proxy as simple as possible give us less attack surface against ennemies exterior to the cozy

4. & 6 TBD, but i envision for instance a service allowing to get some contacts as vcard or some events as ical (ie. make contact & calendar clientsidable). Or a service exposing a git-compatible API, so one can push-deploy to one's cozy.
Basically, I agree the default should be REST + JSON, but let's be open to other opportunities and not make this part of the design document as it could cause to make blocking decisions.
Regarding API format, json-api looks cool and standards are great, but its a bit complicated to handle. I am partial to HTTP-REST (without HATEOAS) JSON-LD.

File: doc/service.md (hunk 5 above)
Comment (nono) [inline doc/service.md]: PSON is no longer maintained. ProtocolBuffer enforces using a schema. Both are not as ubiquitous as JSON (for example, JSON is displayed nicely in chrome and firefox devtools, you can read it when using curl, you can use it with almost any programming language). And for performances, it's not clear if the download size (> 10ko) is not larger than the bytes transfered on calls. Maybe, JSON with brotli compression is even smaller than ProtocolBuffer.

**Comment by nono** [comment]
About json-api vs json-ld vs HAL vs collection+json vs ..., other opinions?

File: doc/service.md (hunk 5 above)
Comment (m4dz) [inline doc/service.md]: > Maybe, JSON with brotli compression

:+1:, sounds definitely good for me

**Comment by frankrousseau** [comment]
Our Twitter community voted for json-api (65%, 25 votes)

**Comment by aenario** [comment]
One addition I just thought about : 

It would make sense to separate the notion of service and the notion of permission.
i.e. there should be no code in the service to handle permission.

Instead the service should pass the authentication from the request to an inner API, which would be responsible for checking permissions.

This would allow for easier review of services security.

**Comment by nono** [comment]
@aenario can you explain what you have in mind? Is it something like a `hasPermission(req.authenticationStuffFromMiddlewares, permissionScope)` where `permissionScope` describe what is the scope needed (like reading images or installing a new app)?

**Comment by frankrousseau** [comment]
I'm ok with the RFC. I just don't like much the use of `I` to describe something. It makes it more human but it's like it's not decided yet. Well, the goal is to find a consensus on how to write properly services.

**Comment by frankrousseau** [comment]
@nono, @aenario can I merge it?

**Comment by nono** [comment]
@frankrousseau I've rewritten the parts with `I`, now that it is a more collective thing. You can merge it.

**Comment by frankrousseau** [comment]
Thank you @nono for that great work. It makes things really clearer about that notion of services. Thank you @aenario and @jsilvestre for the reviews.
